pulp/apis/sas_api.py

```diff
@@ -0,0 +1,623 @@
+                                option notes nonumber nodate nosource pagesize=max;
+                                {warmstart}
+                                %MPS2SASD(MPSFILE="{mpsfile}", OUTDATA=mpsdata{postfix}, MAXLEN={maxLen}, FORMAT=FREE);
+                                proc {proc} data=mpsdata{postfix} {options} primalout=primalout{postfix} dualout=dualout{postfix};
+                                {decomp}
+                                {decompmaster}
+                                {decompmasterip}
+                                {decompsubprob}
+                                {rootnode}
+                                proc delete data=mpsdata{postfix};
+                                run;
+                                """.format(
+                        warmstart=warmstart_str,
+                        postfix=postfix,
+                        mpsfile=tmpMps,
+                        proc=proc,
+                        maxLen=min(nameLen, MAX_NAME_LENGTH),
+                        options=opt_str,
+                        decomp=decomp_str,
+                        decompmaster=decompmaster_str,
+                        decompmasterip=decompmasterip_str,
+                        decompsubprob=decompsubprob_str,
+                        rootnode=rootnode_str,
+                    ),
+                    results="TEXT",
+                )
+            else:
+                # Since 9.4M6+ optlp/optmilp can read mps files directly
+                # TODO: Check whether there are limits for length of variable and constraint names
+                res = sas.submit(
+                    """
+                                option notes nonumber nodate nosource pagesize=max;
+                                {warmstart}
+                                proc {proc} mpsfile=\"{mpsfile}\" {options} primalout=primalout{postfix} dualout=dualout{postfix};
+                                {decomp}
+                                {decompmaster}
+                                {decompmasterip}
+                                {decompsubprob}
+                                {rootnode}
+                                run;
+                                """.format(
+                        warmstart=warmstart_str,
+                        postfix=postfix,
+                        proc=proc,
+                        mpsfile=tmpMps,
+                        options=opt_str,
+                        decomp=decomp_str,
+                        decompmaster=decompmaster_str,
+                        decompmasterip=decompmasterip_str,
+                        decompsubprob=decompsubprob_str,
+                        rootnode=rootnode_str,
+                    ),
+                    results="TEXT",
+                )
+
+            self._delete_tmp_files(tmpMps, tmpMst)
+
+            # Store log and ODS output
+            self._log = res["LOG"]
+            if self.msg:
+                print(self._log)
+            self._macro = dict(
```

Comment:
same goes for `_macro`. is it needed after the solving? why?

Reply:
_macro is used to store some information returned by SAS solvers. It has information like primal and dual infeasibility, and SAS solution status. It might be interesting to SAS users.